src/coordinator/storage/block.go

```diff
@@ -1,58 +1,157 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
 package storage
 
 import (
 	"time"
 
-	"github.com/m3db/m3db/src/coordinator/models"
+	"github.com/m3db/m3db/src/coordinator/block"
 	"github.com/m3db/m3db/src/coordinator/ts"
 )
 
-// Block represents a group of series across a time bound
-type Block interface {
-	Meta() BlockMetadata
-	StepIter() StepIter
-	SeriesIter() SeriesIter
-	SeriesMeta() []SeriesMeta
-	StepMeta() []StepMeta
+// FetchResultToBlockResult converts a fetch result into coordinator blocks
+func FetchResultToBlockResult(result *FetchResult, query *FetchQuery) (block.Result, error) {
+	alignedSeriesList, err := result.SeriesList.Align(query.Start, query.End, query.Interval)
+	if err != nil {
+		return block.Result{}, err
+	}
+
+	multiBlock, err := newMultiSeriesBlock(alignedSeriesList, query)
+	if err != nil {
+		return block.Result{}, err
+	}
+
+	return block.Result{
+		Blocks: []block.Block{multiBlock},
+	}, nil
+}
+
+type multiSeriesBlock struct {
+	seriesList ts.SeriesList
+	meta       block.Metadata
+}
+
+func newMultiSeriesBlock(seriesList ts.SeriesList, query *FetchQuery) (multiSeriesBlock, error) {
+	resolution, err := seriesList.Resolution()
+	if err != nil {
+		return multiSeriesBlock{}, err
+	}
+
+	meta := block.Metadata{
+		Bounds: block.Bounds{
+			Start:    query.Start,
+			End:      query.End,
+			StepSize: resolution,
+		},
+	}
+	return multiSeriesBlock{seriesList: seriesList, meta: meta}, nil
+}
+
+func (m multiSeriesBlock) Meta() block.Metadata {
+	return m.meta
+}
+
+func (m multiSeriesBlock) StepIter() block.StepIter {
+	return &multiSeriesBlockStepIter{block: m}
+}
+
+func (m multiSeriesBlock) SeriesIter() block.SeriesIter {
+	return &multiSeriesBlockSeriesIter{block: m}
 }
 
-// SeriesMeta is metadata data for the series
-type SeriesMeta struct {
-	Tags models.Tags
+func (m multiSeriesBlock) SeriesMeta() []block.SeriesMeta {
+	metas := make([]block.SeriesMeta, len(m.seriesList))
+	for i, s := range m.seriesList {
+		metas[i].Tags = s.Tags
+	}
+
+	return metas
+}
+
+func (m multiSeriesBlock) Close() error {
+	return nil
+}
+
+type multiSeriesBlockStepIter struct {
+	block multiSeriesBlock
+	index int
+}
+
+func (m *multiSeriesBlockStepIter) Next() bool {
```

Comment:
This is a really weird way of moving through the iterator, I'm not super sure how you'd use these iterators... any chance we could get an example test that would illustrate how these are expected to be used?

Reply:
i'll add a few cases

Reply:
CountNode shows the usage of StepIter: src/coordinator/functions/count.go
For multiBlockIter: We need another endpoint which calls FetchBlocks, I'm adding that in another diff.